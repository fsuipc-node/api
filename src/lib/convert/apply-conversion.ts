import { VM } from 'vm2';

import { Offset } from '@shared/offset';
import { RawOffsetValue } from '@shared/offset-values';
import { ConvertedOffsetValue } from '@shared/converted-offset-values';

export const applyConversion = (offset: Offset, rawOffsetValue: RawOffsetValue): ConvertedOffsetValue => {
  if (!offset.convert) {
    return rawOffsetValue;
  }

  if (offset.hasMapping) {
    if (offset.isMappingInvalid) {
      return 'INVALID_MAPPING_FUNCTION';
    }

    return offset.mappingFunction(rawOffsetValue);
  } else {
    // tslint:disable-next-line:no-eval
    if (offset.isInvalidConvertExpression) {
      return 'UNSUPPORTED_CONVERSION_EXPRESSION';
    }

    return new VM().run(
      replaceOffsetExpressionValue(offset, rawOffsetValue)
    );
  }
};

export const replaceOffsetExpressionValue = (offset: Offset, rawOffsetValue: RawOffsetValue): string => {
  return offset.convert.replace(
    new RegExp(/{VAL}/g),
    Array.isArray(rawOffsetValue)
      ? `${JSON.stringify(rawOffsetValue)}`
      : typeof rawOffsetValue === 'string'
        ? `'${rawOffsetValue}'`
        : rawOffsetValue.toString()
  );
};
