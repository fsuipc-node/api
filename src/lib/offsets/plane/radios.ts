import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const radios: OffsetList = {
  comFreq: new Offset({
    value: 0x34E,
    name: 'comFreq',
    category: OffsetCategory.RADIOS,
    description: 'Com frequency',
    type: Type.UInt16,
    convert: 'parseInt(`1` + ({VAL}).toString(16))',
    permission: 'rw',
  }),
  transponderFreq: new Offset({
    value: 0x354,
    name: 'transponderFreq',
    category: OffsetCategory.RADIOS,
    description: 'XPND transponder frequency',
    type: Type.UInt16,
    convert: 'parseInt(`1` + ({VAL}).toString(16))',
    permission: 'rw',
  }),
  nav12Select: new Offset({
    value: 0x374,
    name: 'nav12Select',
    category: OffsetCategory.RADIOS,
    description: 'NAV1/NAV2 select',
    type: Type.UInt16,
    permission: 'rw',
  }),
  dme12Select: new Offset({
    value: 0x378,
    name: 'dme12Select',
    category: OffsetCategory.RADIOS,
    description: 'DME1/DME2 select',
    type: Type.UInt16,
    permission: 'rw',
  }),
  navAdfActivate: new Offset({
    value: 0x388,
    name: 'navAdfActivate',
    category: OffsetCategory.RADIOS,
    description: 'NAV and ADF activate < FS2000',
    type: Type.UInt16,
    permission: 'w',
  }),
  comAtisActivate: new Offset({
    value: 0x38A,
    name: 'comAtisActivate',
    category: OffsetCategory.RADIOS,
    description: 'COM/ATIS activate < FS2000',
    type: Type.UInt16,
    permission: 'w',
  }),
  innerMarker: new Offset({
    value: 0xBAC,
    name: 'innerMarker',
    category: OffsetCategory.RADIOS,
    description: 'inner marker - activated when TRUE',
    type: Type.Int16,
    permission: 'r',
  }),
  middleMarker: new Offset({
    value: 0xBAE,
    name: 'middleMarker',
    category: OffsetCategory.RADIOS,
    description: 'middle marker - activated when TRUE',
    type: Type.Int16,
    permission: 'r',
  }),
  outerMarker: new Offset({
    value: 0xBB0,
    name: 'outerMarker',
    category: OffsetCategory.RADIOS,
    description: 'outer marker - activated when TRUE',
    type: Type.Int16,
    permission: 'r',
  }),
};
