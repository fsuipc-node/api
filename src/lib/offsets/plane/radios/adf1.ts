import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const ADF1: OffsetList = {
  adfFreq: new Offset({
    value: 0x34C,
    name: 'adfFreq',
    category: OffsetCategory.RADIOS,
    description: 'ADF frequency show as Binary Coded Decimal. The thousands digit and the fractional parts are provided in adfFreqExtended',
    type: Type.UInt16,
    convert: '+({VAL}).toString(16)',
    permission: 'rw',
  }),
  adfFreqExtended: new Offset({
    value: 0x356,
    name: 'adfFreqExtended',
    category: OffsetCategory.RADIOS,
    description: 'ADF frequency extended',
    type: Type.UInt16,
    permission: 'rw',
  }),
  adf1SignalStrength: new Offset({
    value: 0xC1C,
    name: 'adf1SignalStrength',
    category: OffsetCategory.RADIOS,
    description: 'ADF1 signal strength',
    type: Type.UInt32,
    permission: 'r',
  }),
  adf1BearingToNDB: new Offset({
    value: 0xC6A,
    name: 'adf1BearingToNDB',
    category: OffsetCategory.RADIOS,
    description: 'ADF1 relative bearing to NDB - degrees - -ve left, +ve right',
    convert: 'Math.round({VAL} * 360 / 65536)',
    type: Type.Int16,
    permission: 'r',
  }),
  adf1DialBearing: new Offset({
    value: 0xC6C,
    name: 'adf1DialBearing',
    category: OffsetCategory.RADIOS,
    description: 'ADF1 dial bearing - degrees',
    type: Type.UInt16,
    permission: 'rw',
  }),
};
