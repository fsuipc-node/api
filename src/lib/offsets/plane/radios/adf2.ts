import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const ADF2: OffsetList = {
  adf2Freq: new Offset({
    value: 0x2D4,
    name: 'adf2Freq',
    category: OffsetCategory.RADIOS,
    description: 'ADF 2 freq - Main 3 digits in BCD - FS2004',
    type: Type.UInt16,
    convert: '+({VAL}).toString(16)',
    permission: 'rw',
  }),
  adf2ExtendedFreq: new Offset({
    value: 0x2D6,
    name: 'adf2ExtendedFreq',
    category: OffsetCategory.RADIOS,
    description: 'ADF 2 extended freq - in BCD - FS2004',
    type: Type.UInt16,
    permission: 'rw',
  }),
  adf2RelBearing: new Offset({
    value: 0x2D8,
    name: 'adf2RelBearing',
    category: OffsetCategory.RADIOS,
    description: 'ADF2 Rel Bearing - FS2004',
    convert: 'Math.round({VAL} * 360 / 65536)',
    type: Type.Int16,
    permission: 'r',
  }),
  ndb2Identity: new Offset({
    value: 0x2DC,
    name: 'ndb2Identity',
    category: OffsetCategory.RADIOS,
    description: 'NDB2 identity - FS2004',
    type: Type.String,
    length: 6,
    permission: 'r',
  }),
  ndb2Name: new Offset({
    value: 0x2E2,
    name: 'ndb2Name',
    category: OffsetCategory.RADIOS,
    description: 'NDB2 name - FS2004',
    type: Type.String,
    length: 25,
    permission: 'r',
  }),
  ndb2IdentSoundSwitch: new Offset({
    value: 0x2FB,
    name: 'ndb2IdentSoundSwitch',
    category: OffsetCategory.RADIOS,
    description: 'NDB2 ident sound switch - FS2004',
    type: Type.Byte,
    convert: '!!{VAL}',
    permission: 'rw',
  }),
  adf2SignalStrength: new Offset({
    value: 0xC14,
    name: 'adf2SignalStrength',
    category: OffsetCategory.RADIOS,
    description: 'ADF2 signal strength',
    type: Type.UInt32,
    permission: 'r',
  }),
};
