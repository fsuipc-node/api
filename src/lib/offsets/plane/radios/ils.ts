import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const ILS: OffsetList = {
  innerMarkerLatitude: new Offset({
    value: 0x1100,
    name: 'innerMarkerLatitude',
    category: OffsetCategory.RADIOS,
    description: 'inner marker latitude',
    type: Type.Int32,
    convert: '{VAL} * 90 / 10001750',
    permission: 'r',
  }),
  innerMarkerLongitude: new Offset({
    value: 0x1104,
    name: 'innerMarkerLongitude',
    category: OffsetCategory.RADIOS,
    description: 'inner marker longitude',
    type: Type.Int32,
    convert: '{VAL} * 360 / (65536 * 65536)',
    permission: 'r',
  }),
  innerMarkerAltitude: new Offset({
    value: 0x1108,
    name: 'innerMarkerAltitude',
    category: OffsetCategory.RADIOS,
    description: 'inner marker altitude - ft',
    type: Type.Int32,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  middleMarkerLatitude: new Offset({
    value: 0x110C,
    name: 'middleMarkerLatitude',
    category: OffsetCategory.RADIOS,
    description: 'middle marker latitude',
    type: Type.Int32,
    convert: '{VAL} * 90 / 10001750',
    permission: 'r',
  }),
  middleMarkerLongitude: new Offset({
    value: 0x1110,
    name: 'middleMarkerLongitude',
    category: OffsetCategory.RADIOS,
    description: 'middle marker longitude',
    type: Type.Int32,
    convert: '{VAL} * 360 / (65536 * 65536)',
    permission: 'r',
  }),
  middleMarkerAltitude: new Offset({
    value: 0x1114,
    name: 'middleMarkerAltitude',
    category: OffsetCategory.RADIOS,
    description: 'middle marker altitude - ft',
    type: Type.Int32,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  outerMarkerLatitude: new Offset({
    value: 0x1118,
    name: 'outerMarkerLatitude',
    category: OffsetCategory.RADIOS,
    description: 'outer marker latitude',
    type: Type.Int32,
    convert: '{VAL} * 90 / 10001750',
    permission: 'r',
  }),
  outerMarkerLongitude: new Offset({
    value: 0x111C,
    name: 'outerMarkerLongitude',
    category: OffsetCategory.RADIOS,
    description: 'outer marker longitude',
    type: Type.Int32,
    convert: '{VAL} * 360 / (65536 * 65536)',
    permission: 'r',
  }),
  outerMarkerAltitude: new Offset({
    value: 0x1120,
    name: 'outerMarkerAltitude',
    category: OffsetCategory.RADIOS,
    description: 'outer marker altitude - ft',
    type: Type.Int32,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
};
