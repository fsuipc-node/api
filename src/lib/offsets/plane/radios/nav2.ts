import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const NAV2: OffsetList = {
  vor2DMEDistance: new Offset({
    value: 0x306,
    name: 'vor2DMEDistance',
    category: OffsetCategory.RADIOS,
    description: 'VOR2 DME distance - nm',
    convert: '{VAL} / 10',
    type: Type.UInt16,
    permission: 'r',
  }),
  vor2DMESpeed: new Offset({
    value: 0x308,
    name: 'vor2DMESpeed',
    category: OffsetCategory.RADIOS,
    description: 'VOR2 DME speed - kt',
    convert: '{VAL} / 10',
    type: Type.UInt16,
    permission: 'r',
  }),
  vor2DMETimeToStation: new Offset({
    value: 0x30A,
    name: 'vor2DMETimeToStation',
    category: OffsetCategory.RADIOS,
    description: 'VOR2 DME time to station - sec',
    convert: '{VAL} / 10',
    type: Type.UInt16,
    permission: 'r',
  }),
  nav2Freq: new Offset({
    value: 0x352,
    name: 'nav2Freq',
    category: OffsetCategory.RADIOS,
    description: 'NAV2 frequency',
    type: Type.UInt16,
    convert: 'parseInt(`1` + ({VAL}).toString(16))',
    permission: 'rw',
  }),
  hasNav2: new Offset({
    value: 0x7A4,
    name: 'hasNav2',
    category: OffsetCategory.RADIOS,
    description: 'has NAV2',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  vor2DmeLatitude: new Offset({
    value: 0x834,
    name: 'vor2DmeLatitude',
    category: OffsetCategory.RADIOS,
    description: 'VOR 2 DME latitude - FS2002+',
    type: Type.Int32,
    convert: '{VAL} * 90 / 10001750',
    permission: 'r',
  }),
  vor2DmeLongitude: new Offset({
    value: 0x838,
    name: 'vor2DmeLongitude',
    category: OffsetCategory.RADIOS,
    description: 'VOR 2 DME longitude - FS2002+',
    type: Type.Int32,
    convert: '{VAL} * 360 / (65536 * 65536)',
    permission: 'r',
  }),
  vor2DmeElevation: new Offset({
    value: 0x83C,
    name: 'vor2DmeElevation',
    category: OffsetCategory.RADIOS,
    description: 'VOR 2 DME elevation - meters - FS2002+',
    type: Type.Int32,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  vor2ILSLocHeadingTrue: new Offset({
    value: 0x844,
    name: 'vor2ILSLocHeadingTrue',
    category: OffsetCategory.RADIOS,
    description: 'NAV2 ILS Localiser inverse runway heading if VOR 2 is ILS - TRUE - FS2002+ - 180 different to aircraft direction to follow localiser',
    type: Type.UInt16,
    convert: '+({VAL} * 360 / 65536).toFixed(2)',
    permission: 'r',
  }),
  vor2ILSGlideSlopeAngle: new Offset({
    value: 0x846,
    name: 'vor2ILSGlideSlopeAngle',
    category: OffsetCategory.RADIOS,
    description: 'VOR 2 ILS GlideSlope Angle',
    type: Type.Int16,
    convert: '+({VAL} * 360 / 65536).toFixed(2)',
    permission: 'r',
  }),
  vor2LocLatitude: new Offset({
    value: 0x84C,
    name: 'vor2LocLatitude',
    category: OffsetCategory.RADIOS,
    description: 'VOR 2 or NAV2 ILS LOC latitude - FS2002+',
    type: Type.Int32,
    convert: '{VAL} * 90 / 10001750',
    permission: 'r',
  }),
  vor2LocLongitude: new Offset({
    value: 0x850,
    name: 'vor2LocLongitude',
    category: OffsetCategory.RADIOS,
    description: 'VOR 2 or NAV2 ILS LOC longitude - FS2002+',
    type: Type.Int32,
    convert: '{VAL} * 360 / (65536 * 65536)',
    permission: 'r',
  }),
  vor2LocElevation: new Offset({
    value: 0x854,
    name: 'vor2LocElevation',
    category: OffsetCategory.RADIOS,
    description: 'VOR 2 or NAV2 ILS LOC elevation - meters - FS2002+',
    type: Type.Int32,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  vor2OrILSGlideSlopeLatitude: new Offset({
    value: 0x858,
    name: 'vor2OrILSGlideSlopeLatitude',
    category: OffsetCategory.RADIOS,
    description: 'VOR 2 latitude or NAV2 ILS glideslope latitude - FS2002+',
    type: Type.Int32,
    convert: '{VAL} * 90 / 10001750',
    permission: 'r',
  }),
  vor2OrILSGlideSlopeLongitude: new Offset({
    value: 0x860,
    name: 'vor2OrILSGlideSlopeLongitude',
    category: OffsetCategory.RADIOS,
    description: 'VOR 2 or NAV 2 ILS glideslope longitude - FS2002+',
    type: Type.Int32,
    convert: '{VAL} * 360 / (65536 * 65536)',
    permission: 'r',
  }),
  vor2OrILSGlideSlopeElevation: new Offset({
    value: 0x868,
    name: 'vor2OrILSGlideSlopeElevation',
    category: OffsetCategory.RADIOS,
    description: 'VOR 2 or NAV2 ILS glideslope elevation - meters - FS2002+',
    type: Type.Int32,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  dme2Distance: new Offset({
    value: 0xC33,
    name: 'dme2Distance',
    category: OffsetCategory.RADIOS,
    description: 'DME2 distance - nm',
    convert: '+{VAL}',
    type: Type.String,
    length: 5,
    permission: 'r',
  }),
  dme2Speed: new Offset({
    value: 0xC38,
    name: 'dme2Speed',
    category: OffsetCategory.RADIOS,
    description: 'DME2 speed - kt',
    convert: '+{VAL}',
    type: Type.String,
    length: 5,
    permission: 'r',
  }),
  nav2MagVar: new Offset({
    value: 0xC42,
    name: 'nav2MagVar',
    category: OffsetCategory.RADIOS,
    description: 'NAV2 mag var',
    convert: 'Math.round({VAL} * 360 / 65536)',
    type: Type.Int16,
    permission: 'r',
  }),
  nav2LocNeedlePosition: new Offset({
    value: 0xC59,
    name: 'nav2LocNeedlePosition',
    category: OffsetCategory.RADIOS,
    description: 'NAV2 loc needle position - percent - -ve left +ve right',
    convert: 'Math.round({VAL} / 127 * 100)',
    type: Type.SByte,
    permission: 'r',
  }),
  nav2BackCourseFlags: new Offset({
    value: 0xC5A,
    name: 'nav2BackCourseFlags',
    category: OffsetCategory.RADIOS,
    description: 'NAV2 glideslope flags',
    convert: 'navBackCourseFlags',
    mapping: true,
    type: Type.ByteArray,
    length: 8,
    permission: 'r',
  }),
  nav2ToFrom: new Offset({
    value: 0xC5B,
    name: 'nav2ToFrom',
    category: OffsetCategory.RADIOS,
    description: 'NAV2 TO/FROM flag - values: off/to/from',
    convert: 'vorToFrom',
    mapping: true,
    type: Type.Byte,
    permission: 'r',
  }),
  nav2BearingToVOR: new Offset({
    value: 0xC5C,
    name: 'nav2BearingToVOR',
    category: OffsetCategory.RADIOS,
    description: 'NAV2 relative bearing to VOR - degrees',
    type: Type.UInt16,
    permission: 'r',
  }),
  nav2Obs: new Offset({
    value: 0xC5E,
    name: 'nav2Obs',
    category: OffsetCategory.RADIOS,
    description: 'NAV2 obs setting - degrees',
    type: Type.Int16,
    permission: 'rw',
  }),
  nav2Radial: new Offset({
    value: 0xC60,
    name: 'nav2Radial',
    category: OffsetCategory.RADIOS,
    description: 'NAV2 radial - degrees - MAG for VOR, TRUE for ILS LOC',
    convert: 'Math.round({VAL} * 360 / 65536)',
    type: Type.UInt16,
    permission: 'r',
  }),
  nav2SignalStrength: new Offset({
    value: 0xC62,
    name: 'nav2SignalStrength',
    category: OffsetCategory.RADIOS,
    description: 'NAV2 signal strength',
    type: Type.UInt32,
    permission: 'r',
  }),
  nav2GlideslopeNeedlePosition: new Offset({
    value: 0xC6E,
    name: 'nav2GlideslopeNeedlePosition',
    category: OffsetCategory.RADIOS,
    description: 'NAV2 glideslope needle position - percent - -ve up +ve down',
    convert: 'Math.round({VAL} / 127 * 100)',
    type: Type.SByte,
    permission: 'r',
  }),
  nav2GlideslopeActive: new Offset({
    value: 0xC6F,
    name: 'nav2GlideslopeActive',
    category: OffsetCategory.RADIOS,
    description: 'NAV2 glideslope flag',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  nav2Capabilities: new Offset({
    value: 0xC70,
    name: 'nav2Capabilities',
    category: OffsetCategory.RADIOS,
    description: 'NAV2 capabilities',
    convert: 'navCapabilities',
    mapping: true,
    type: Type.ByteArray,
    length: 8,
    permission: 'r',
  }),
};
