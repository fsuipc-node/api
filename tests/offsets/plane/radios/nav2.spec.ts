import { NAV2 as offsets } from '@offsets/plane/radios/nav2';
import { replaceOffsetExpressionValue } from '@convert/apply-conversion';

describe('offset - plane/radios/nav2', () => {
  const offsetsTestCases = [
    { name: 'vor2DMEDistance', value: 550, expectedResult: 55 },
    { name: 'vor2DMESpeed', value: 650, expectedResult: 65 },
    { name: 'vor2DMETimeToStation', value: 750, expectedResult: 75 },
    { name: 'nav2Freq', value: 13584, expectedResult: 13510 },
    { name: 'hasNav2', value: 0, expectedResult: false },
    { name: 'hasNav2', value: 1, expectedResult: true },
    { name: 'vor2DmeLatitude', value: 5198687, expectedResult: 46.77999650061239 },
    { name: 'vor2DmeLongitude', value: -851715875, expectedResult: -71.38999993912876 },
    { name: 'vor2DmeElevation', value: 92, expectedResult: 301.84 },
    { name: 'vor2ILSLocHeadingTrue', value: 27346, expectedResult: 150.22 },
    { name: 'vor2ILSGlideSlopeAngle', value: 27346, expectedResult: 150.22 },
    { name: 'vor2LocLatitude', value: 5198687, expectedResult: 46.77999650061239 },
    { name: 'vor2LocLongitude', value: -851715875, expectedResult: -71.38999993912876 },
    { name: 'vor2LocElevation', value: 92, expectedResult: 301.84 },
    { name: 'vor2OrILSGlideSlopeLatitude', value: 5198687, expectedResult: 46.77999650061239 },
    { name: 'vor2OrILSGlideSlopeLongitude', value: -851715875, expectedResult: -71.38999993912876 },
    { name: 'vor2OrILSGlideSlopeElevation', value: 92, expectedResult: 301.84 },
    { name: 'dme2Distance', value: '92.5', expectedResult: 92.5 },
    { name: 'dme2Distance', value: '92.5 ', expectedResult: 92.5 },
    { name: 'dme2Distance', value: '92.50', expectedResult: 92.5 },
    { name: 'dme2Distance', value: '112.5', expectedResult: 112.5 },
    { name: 'dme2Distance', value: '112.0', expectedResult: 112 },
    { name: 'dme2Distance', value: '112. ', expectedResult: 112 },
    { name: 'dme2Speed', value: '92.5', expectedResult: 92.5 },
    { name: 'dme2Speed', value: '92.5 ', expectedResult: 92.5 },
    { name: 'dme2Speed', value: '92.50', expectedResult: 92.5 },
    { name: 'dme2Speed', value: '112.5', expectedResult: 112.5 },
    { name: 'dme2Speed', value: '112.0', expectedResult: 112 },
    { name: 'dme2Speed', value: '112. ', expectedResult: 112 },
    { name: 'nav2MagVar', value: 9102, expectedResult: 50 },
    { name: 'nav2LocNeedlePosition', value: 127, expectedResult: 100 },
    { name: 'nav2LocNeedlePosition', value: -127, expectedResult: -100 },
    { name: 'nav2Radial', value: 9102, expectedResult: 50 },
    { name: 'nav2GlideslopeNeedlePosition', value: 127, expectedResult: 100 },
    { name: 'nav2GlideslopeNeedlePosition', value: -127, expectedResult: -100 },
    { name: 'nav2GlideslopeActive', value: 0, expectedResult: false },
    { name: 'nav2GlideslopeActive', value: 1, expectedResult: true },
  ];

  describe('offsets list', () => {
    it('should have required properties', () => {
      expect(offsets).toMatchSnapshot();
    });
  });

  offsetsTestCases.forEach(testedOffset => {
    describe(testedOffset.name, () => {
      it('should convert data properly', () => {
        const convertExpression = replaceOffsetExpressionValue(offsets[testedOffset.name], testedOffset.value);

        // tslint:disable-next-line:no-eval
        expect(eval(convertExpression)).toEqual(testedOffset.expectedResult);
      });
    });
  });
});
