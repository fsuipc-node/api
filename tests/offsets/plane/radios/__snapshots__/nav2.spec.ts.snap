// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offset - plane/radios/nav2 offsets list should have required properties 1`] = `
Object {
  "dme2Distance": Offset {
    "category": "radios",
    "convert": "+{VAL}",
    "description": "DME2 distance - nm",
    "length": 5,
    "mapping": undefined,
    "name": "dme2Distance",
    "permission": "r",
    "type": 11,
    "value": 3123,
  },
  "dme2Speed": Offset {
    "category": "radios",
    "convert": "+{VAL}",
    "description": "DME2 speed - kt",
    "length": 5,
    "mapping": undefined,
    "name": "dme2Speed",
    "permission": "r",
    "type": 11,
    "value": 3128,
  },
  "hasNav2": Offset {
    "category": "radios",
    "convert": "!!{VAL}",
    "description": "has NAV2",
    "length": undefined,
    "mapping": undefined,
    "name": "hasNav2",
    "permission": "r",
    "type": 6,
    "value": 1956,
  },
  "nav2BackCourseFlags": Offset {
    "category": "radios",
    "convert": "navBackCourseFlags",
    "description": "NAV2 glideslope flags",
    "length": 8,
    "mapping": true,
    "name": "nav2BackCourseFlags",
    "permission": "r",
    "type": 10,
    "value": 3162,
  },
  "nav2BearingToVOR": Offset {
    "category": "radios",
    "convert": undefined,
    "description": "NAV2 relative bearing to VOR - degrees",
    "length": undefined,
    "mapping": undefined,
    "name": "nav2BearingToVOR",
    "permission": "r",
    "type": 5,
    "value": 3164,
  },
  "nav2Capabilities": Offset {
    "category": "radios",
    "convert": "navCapabilities",
    "description": "NAV2 capabilities",
    "length": 8,
    "mapping": true,
    "name": "nav2Capabilities",
    "permission": "r",
    "type": 10,
    "value": 3184,
  },
  "nav2Freq": Offset {
    "category": "radios",
    "convert": "parseInt(\`1\` + ({VAL}).toString(16))",
    "description": "NAV2 frequency",
    "length": undefined,
    "mapping": undefined,
    "name": "nav2Freq",
    "permission": "rw",
    "type": 5,
    "value": 850,
  },
  "nav2GlideslopeActive": Offset {
    "category": "radios",
    "convert": "!!{VAL}",
    "description": "NAV2 glideslope flag",
    "length": undefined,
    "mapping": undefined,
    "name": "nav2GlideslopeActive",
    "permission": "r",
    "type": 0,
    "value": 3183,
  },
  "nav2GlideslopeNeedlePosition": Offset {
    "category": "radios",
    "convert": "Math.round({VAL} / 127 * 100)",
    "description": "NAV2 glideslope needle position - percent - -ve up +ve down",
    "length": undefined,
    "mapping": undefined,
    "name": "nav2GlideslopeNeedlePosition",
    "permission": "r",
    "type": 1,
    "value": 3182,
  },
  "nav2LocNeedlePosition": Offset {
    "category": "radios",
    "convert": "Math.round({VAL} / 127 * 100)",
    "description": "NAV2 loc needle position - percent - -ve left +ve right",
    "length": undefined,
    "mapping": undefined,
    "name": "nav2LocNeedlePosition",
    "permission": "r",
    "type": 1,
    "value": 3161,
  },
  "nav2MagVar": Offset {
    "category": "radios",
    "convert": "Math.round({VAL} * 360 / 65536)",
    "description": "NAV2 mag var",
    "length": undefined,
    "mapping": undefined,
    "name": "nav2MagVar",
    "permission": "r",
    "type": 2,
    "value": 3138,
  },
  "nav2Obs": Offset {
    "category": "radios",
    "convert": undefined,
    "description": "NAV2 obs setting - degrees",
    "length": undefined,
    "mapping": undefined,
    "name": "nav2Obs",
    "permission": "rw",
    "type": 2,
    "value": 3166,
  },
  "nav2Radial": Offset {
    "category": "radios",
    "convert": "Math.round({VAL} * 360 / 65536)",
    "description": "NAV2 radial - degrees - MAG for VOR, TRUE for ILS LOC",
    "length": undefined,
    "mapping": undefined,
    "name": "nav2Radial",
    "permission": "r",
    "type": 5,
    "value": 3168,
  },
  "nav2SignalStrength": Offset {
    "category": "radios",
    "convert": undefined,
    "description": "NAV2 signal strength",
    "length": undefined,
    "mapping": undefined,
    "name": "nav2SignalStrength",
    "permission": "r",
    "type": 6,
    "value": 3170,
  },
  "nav2ToFrom": Offset {
    "category": "radios",
    "convert": "vorToFrom",
    "description": "NAV2 TO/FROM flag - values: off/to/from",
    "length": undefined,
    "mapping": true,
    "name": "nav2ToFrom",
    "permission": "r",
    "type": 0,
    "value": 3163,
  },
  "vor2DMEDistance": Offset {
    "category": "radios",
    "convert": "{VAL} / 10",
    "description": "VOR2 DME distance - nm",
    "length": undefined,
    "mapping": undefined,
    "name": "vor2DMEDistance",
    "permission": "r",
    "type": 5,
    "value": 774,
  },
  "vor2DMESpeed": Offset {
    "category": "radios",
    "convert": "{VAL} / 10",
    "description": "VOR2 DME speed - kt",
    "length": undefined,
    "mapping": undefined,
    "name": "vor2DMESpeed",
    "permission": "r",
    "type": 5,
    "value": 776,
  },
  "vor2DMETimeToStation": Offset {
    "category": "radios",
    "convert": "{VAL} / 10",
    "description": "VOR2 DME time to station - sec",
    "length": undefined,
    "mapping": undefined,
    "name": "vor2DMETimeToStation",
    "permission": "r",
    "type": 5,
    "value": 778,
  },
  "vor2DmeElevation": Offset {
    "category": "radios",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "VOR 2 DME elevation - meters - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "vor2DmeElevation",
    "permission": "r",
    "type": 3,
    "value": 2108,
  },
  "vor2DmeLatitude": Offset {
    "category": "radios",
    "convert": "{VAL} * 90 / 10001750",
    "description": "VOR 2 DME latitude - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "vor2DmeLatitude",
    "permission": "r",
    "type": 3,
    "value": 2100,
  },
  "vor2DmeLongitude": Offset {
    "category": "radios",
    "convert": "{VAL} * 360 / (65536 * 65536)",
    "description": "VOR 2 DME longitude - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "vor2DmeLongitude",
    "permission": "r",
    "type": 3,
    "value": 2104,
  },
  "vor2ILSGlideSlopeAngle": Offset {
    "category": "radios",
    "convert": "+({VAL} * 360 / 65536).toFixed(2)",
    "description": "VOR 2 ILS GlideSlope Angle",
    "length": undefined,
    "mapping": undefined,
    "name": "vor2ILSGlideSlopeAngle",
    "permission": "r",
    "type": 2,
    "value": 2118,
  },
  "vor2ILSLocHeadingTrue": Offset {
    "category": "radios",
    "convert": "+({VAL} * 360 / 65536).toFixed(2)",
    "description": "NAV2 ILS Localiser inverse runway heading if VOR 2 is ILS - TRUE - FS2002+ - 180 different to aircraft direction to follow localiser",
    "length": undefined,
    "mapping": undefined,
    "name": "vor2ILSLocHeadingTrue",
    "permission": "r",
    "type": 5,
    "value": 2116,
  },
  "vor2LocElevation": Offset {
    "category": "radios",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "VOR 2 or NAV2 ILS LOC elevation - meters - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "vor2LocElevation",
    "permission": "r",
    "type": 3,
    "value": 2132,
  },
  "vor2LocLatitude": Offset {
    "category": "radios",
    "convert": "{VAL} * 90 / 10001750",
    "description": "VOR 2 or NAV2 ILS LOC latitude - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "vor2LocLatitude",
    "permission": "r",
    "type": 3,
    "value": 2124,
  },
  "vor2LocLongitude": Offset {
    "category": "radios",
    "convert": "{VAL} * 360 / (65536 * 65536)",
    "description": "VOR 2 or NAV2 ILS LOC longitude - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "vor2LocLongitude",
    "permission": "r",
    "type": 3,
    "value": 2128,
  },
  "vor2OrILSGlideSlopeElevation": Offset {
    "category": "radios",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "VOR 2 or NAV2 ILS glideslope elevation - meters - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "vor2OrILSGlideSlopeElevation",
    "permission": "r",
    "type": 3,
    "value": 2152,
  },
  "vor2OrILSGlideSlopeLatitude": Offset {
    "category": "radios",
    "convert": "{VAL} * 90 / 10001750",
    "description": "VOR 2 latitude or NAV2 ILS glideslope latitude - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "vor2OrILSGlideSlopeLatitude",
    "permission": "r",
    "type": 3,
    "value": 2136,
  },
  "vor2OrILSGlideSlopeLongitude": Offset {
    "category": "radios",
    "convert": "{VAL} * 360 / (65536 * 65536)",
    "description": "VOR 2 or NAV 2 ILS glideslope longitude - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "vor2OrILSGlideSlopeLongitude",
    "permission": "r",
    "type": 3,
    "value": 2144,
  },
}
`;
